File: apps/api/app/schemas/contract.py
from datetime import date, datetime
from decimal import Decimal
from typing import Literal

from pydantic import BaseModel, Field


ContractStatus = Literal["draft", "active", "expired", "terminated", "completed"]


class ContractBase(BaseModel):
    document_id: str
    contract_number: str | None = Field(default=None, max_length=128)
    contract_title: str | None = Field(default=None, max_length=512)
    supplier_name: str | None = Field(default=None, max_length=255)
    sign_date: date | None = None
    effective_from: date | None = None
    effective_to: date | None = None
    contract_value: Decimal | None = Field(default=None, ge=0)
    currency: str = Field(default="VND", min_length=1, max_length=8)
    status: ContractStatus = "draft"
    notes: str | None = None


class ContractCreateRequest(ContractBase):
    pass


class ContractUpdateRequest(BaseModel):
    contract_number: str | None = Field(default=None, max_length=128)
    contract_title: str | None = Field(default=None, max_length=512)
    supplier_name: str | None = Field(default=None, max_length=255)
    sign_date: date | None = None
    effective_from: date | None = None
    effective_to: date | None = None
    contract_value: Decimal | None = Field(default=None, ge=0)
    currency: str = Field(default="VND", min_length=1, max_length=8)
    status: ContractStatus = "draft"
    notes: str | None = None


class ContractRead(BaseModel):
    id: str
    document_id: str
    document_title: str | None = None
    document_number: str | None = None
    document_status: str | None = None
    contract_number: str | None = None
    contract_title: str | None = None
    supplier_name: str | None = None
    sign_date: date | None = None
    effective_from: date | None = None
    effective_to: date | None = None
    contract_value: Decimal | None = None
    currency: str
    status: str
    notes: str | None = None
    created_at: datetime
    updated_at: datetime


class ContractListResponse(BaseModel):
    items: list[ContractRead]
    total: int
    limit: int
    offset: int

File: apps/web/types/contract.ts
export type ContractStatus = 'draft' | 'active' | 'expired' | 'terminated' | 'completed'

export interface ContractItem {
  id: string
  document_id: string
  document_title?: string | null
  document_number?: string | null
  document_status?: string | null
  contract_number?: string | null
  contract_title?: string | null
  supplier_name?: string | null
  sign_date?: string | null
  effective_from?: string | null
  effective_to?: string | null
  contract_value?: string | number | null
  currency: string
  status: ContractStatus
  notes?: string | null
  created_at: string
  updated_at: string
}

export interface ContractInput {
  document_id: string
  contract_number?: string
  contract_title?: string
  supplier_name?: string
  sign_date?: string
  effective_from?: string
  effective_to?: string
  contract_value?: string
  currency: string
  status: ContractStatus
  notes?: string
}

export interface ContractListFilters {
  q?: string
  document_id?: string
  contract_number?: string
  supplier_name?: string
  status?: ContractStatus | ''
  sign_date_from?: string
  sign_date_to?: string
  effective_to_from?: string
  effective_to_to?: string
  sort_by?: 'created_at' | 'updated_at' | 'contract_number' | 'supplier_name' | 'status' | 'sign_date' | 'effective_to'
  sort_dir?: 'asc' | 'desc'
  limit?: number
  offset?: number
}

export interface ContractListResponse {
  items: ContractItem[]
  total: number
  limit: number
  offset: number
}
